substrate/frame/contracts/src/transient_storage.rs

```diff
@@ -0,0 +1,626 @@
+		storage.write(&self.account, &self.key, self.prev_value);
+	}
+}
+
+/// A journal containing transient storage modifications.
+struct Journal(Vec<JournalEntry>);
+
+impl Journal {
+	/// Create a new journal.
+	pub fn new() -> Self {
+		Self(Default::default())
+	}
+
+	/// Add a chenge to the journal.
+	pub fn push(&mut self, entry: JournalEntry) {
+		self.0.push(entry);
+	}
+
+	/// Length of the journal.
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	/// Roll back all journal changes until the chackpoint
+	pub fn rollback(&mut self, storage: &mut Storage, checkpoint: usize) {
+		self.0.drain(checkpoint..).rev().for_each(|entry| entry.revert(storage));
+	}
+}
+
+/// Storage for maintaining the current transaction state.
+#[derive(Default)]
+struct Storage(BTreeMap<Vec<u8>, Vec<u8>>);
+
+impl Storage {
+	/// Read the storage entry.
+	pub fn read(&self, account: &[u8], key: &[u8]) -> Option<Vec<u8>> {
+		self.0.get(&Self::storage_key(account, key)).cloned()
+	}
+
+	/// Write the storage entry.
+	pub fn write(&mut self, account: &[u8], key: &[u8], value: Option<Vec<u8>>) -> Option<Vec<u8>> {
+		if let Some(value) = value {
+			// Insert storage entry.
+			self.0.insert(Self::storage_key(account, key), value)
+		} else {
+			// Remove storage entry.
+			self.0.remove(&Self::storage_key(account, key))
+		}
+	}
+
+	/// Get the storage keys for the account.
+	pub fn keys<'a>(&'a self, account: &'a [u8]) -> impl Iterator<Item = Vec<u8>> + 'a {
+		self.0
+			.range((Included(account.to_vec()), Unbounded))
+			.take_while(|(key, _)| key.starts_with(account))
+			.map(|(key, _)| key[account.len()..].to_vec())
+	}
+
+	fn storage_key(account: &[u8], key: &[u8]) -> Vec<u8> {
+		let mut storage_key = Vec::with_capacity(account.len() + key.len());
+		storage_key.extend_from_slice(&account);
+		storage_key.extend_from_slice(&key);
+		storage_key
+	}
+}
+
+/// Transient storage behaves almost identically to regular storage but is discarded after each
+/// transaction. It consists of a `BTreeMap` for the current state, a journal of all changes, and a
+/// list of checkpoints. On entry to the `start_transaction` function, a marker (checkpoint) is
+/// added to the list. New values are written to the current state, and the previous value is
+/// recorded in the journal (`write`). When the `commit_transaction` function is called, the marker
+/// to the journal index (checkpoint) of when that call was entered is discarded.
+/// On `rollback_transaction`, all entries are reverted up to the last checkpoint.
+pub struct TransientStorage<T: Config> {
+	// The storage and journal size is limited by the storage meter.
+	current: Storage,
+	journal: Journal,
+	// The size of the StorageMeter is limited by the stack depth.
+	meter: StorageMeter<T>,
+	// The size of the checkpoints is limited by the stack depth.
+	checkpoints: Checkpoints,
+}
+
+impl<T: Config> TransientStorage<T> {
+	pub fn new(total_limit: u32, transaction_limit: u32) -> Self {
```

Comment:
I suggest passing in the max call depth size instead of the transaction limit. It makes sense to encapsulate the logic here. That also means removing the default derive.